nat-lab/tests/utils/connection_util.py

```
@@ -492,3 +497,40 @@ async def remove_traffic_control_rules(connection):
         ]).execute()
     except:
         pass
+
+
+async def setup_ephemeral_ports(connection: Connection, connection_tag: ConnectionTag):
+    async def on_output(output: str) -> None:
+        print(datetime.now(), f"[{connection_tag.name}]: {output}")
+
+    start_port = random.randint(5000, 55000)
+    num_ports = random.randint(2000, 5000)
+
+    if connection_tag in [ConnectionTag.WINDOWS_VM_1, ConnectionTag.WINDOWS_VM_2]:
+        cmd = [
+            "netsh",
+            "int",
+            "ipv4",
+            "set",
+            "dynamic",
+            "tcp",
+            f"start={start_port}",
+            f"num={num_ports}",
+        ]
+    elif connection_tag is ConnectionTag.MAC_VM:
+        cmd = [
+            "sysctl",
+            "-w",
+            f"net.inet.ip.portrange.first={start_port}",
+            f"net.inet.ip.portrange.last={start_port + num_ports}",
+        ]
+    elif connection_tag.name.startswith("DOCKER_") and "CLIENT" in connection_tag.name:
+        cmd = [
+            "sysctl",
+            "-w",
+            f"net.ipv4.ip_local_port_range={start_port} {start_port + num_ports}",
+        ]
+    else:
+        return
```

this should throw

---

It shouldn't. Eg: if you call `new_connection_raw()` on a gw container it shouldn't setup the ephemeral ports, in that case it should only return/pass.

---

We only want to setup the port ranges for those 3 cases, the others we can ignore.

---

I would say we don't expect that there is any other case, right? And if that is the case, it's better to loudly signal that something unexpected happened instead of silently ignoring it.

---

Unless this now happens for all docker containers, including those which aren't clients?

---

> I would say we don't expect that there is any other case, right? And if that is the case, it's better to loudly signal that something unexpected happened instead of silently ignoring it.

I agree with that.

> Unless this now happens for all docker containers, including those which aren't clients?

It can happen that someone or some test creates a connection to a container that's not a win/mac VM nor a docker client, like `DOCKER_VPN_1` [here](https://github.com/NordSecurity/libtelio/blob/main/nat-lab/tests/test_pmtu.py#L199).